extensions/azurecore/src/account-provider/auths/azureAuthCodeGrant.ts

```diff
@@ -0,0 +1,230 @@
+
+import * as azdata from 'azdata';
+import * as vscode from 'vscode';
+import * as crypto from 'crypto';
+import * as nls from 'vscode-nls';
+import { promises as fs } from 'fs';
+import * as path from 'path';
+import * as http from 'http';
+import * as qs from 'qs';
+
+import {
+	AzureAuth,
+	AccessToken,
+	RefreshToken,
+	TokenClaims,
+	TokenRefreshResponse,
+} from './azureAuth';
+
+import {
+	AzureAccountProviderMetadata,
+	AzureAuthType,
+	Deferred
+} from '../interfaces';
+
+import { SimpleWebServer } from '../utils/simpleWebServer';
+import { SimpleTokenCache } from '../simpleTokenCache';
+const localize = nls.loadMessageBundle();
+
+export class AzureAuthCodeGrant extends AzureAuth {
+	private static readonly USER_FRIENDLY_NAME: string = localize('azure.azureAuthCodeGrantName', "Azure Auth Code Grant");
+	private server: SimpleWebServer;
+
+	constructor(metadata: AzureAccountProviderMetadata,
+		tokenCache: SimpleTokenCache,
+		context: vscode.ExtensionContext) {
+		super(metadata, tokenCache, context, AzureAuthType.AuthCodeGrant, AzureAuthCodeGrant.USER_FRIENDLY_NAME);
+	}
+
+	public async autoOAuthCancelled(): Promise<void> {
+		return this.server.shutdown();
+	}
+
+	public async login(): Promise<azdata.Account | azdata.PromptFailedResult> {
+		let authCompleteDeferred: Deferred<void>;
+		let authCompletePromise = new Promise<void>((resolve, reject) => authCompleteDeferred = { resolve, reject });
+
+		this.server = new SimpleWebServer();
+		const nonce = crypto.randomBytes(16).toString('base64');
+		let serverPort: string;
+
+		try {
+			serverPort = await this.server.startup();
+		} catch (err) {
+			const msg = localize('azure.serverCouldNotStart', 'Server could not start. This could be a permissions error or an incompatibility on your system. You can try enabling device code authentication from settings.');
+			await vscode.window.showErrorMessage(msg);
+			console.dir(err);
+			return { canceled: false } as azdata.PromptFailedResult;
+		}
+
+
+		// The login code to use
+		let loginUrl: string;
+		let codeVerifier: string;
+		{
+			codeVerifier = this.toBase64UrlEncoding(crypto.randomBytes(32).toString('base64'));
+			const state = `${serverPort},${encodeURIComponent(nonce)}`;
+			const codeChallenge = this.toBase64UrlEncoding(crypto.createHash('sha256').update(codeVerifier).digest('base64'));
+			const loginQuery = {
+				response_type: 'code',
+				response_mode: 'query',
+				client_id: this.clientId,
+				redirect_uri: this.redirectUri,
+				state,
+				prompt: 'select_account',
+				code_challenge_method: 'S256',
+				code_challenge: codeChallenge,
+				resource: this.metadata.settings.signInResourceId
+			};
+			loginUrl = `${this.loginEndpointUrl}${this.commonTenant}/oauth2/authorize?${qs.stringify(loginQuery)}`;
+		}
+
+		await vscode.env.openExternal(vscode.Uri.parse(`http://localhost:${serverPort}/signin?nonce=${encodeURIComponent(nonce)}`));
```

Comment:
So I did bring this back up here because the next one is awaiting for the authenticated code.